src/core/algorithms/pac/model/default_domains/metric_based_domain.cpp

```
@@ -0,0 +1,91 @@
+#include "metric_based_domain.h"
+
+#include <algorithm>
+#include <cstddef>
+#include <iterator>
+#include <limits>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include "algorithms/pac/model/comparable_tuple_type.h"
+#include "algorithms/pac/model/tuple.h"
+#include "exceptions.h"
+
+namespace pac::model {
+using namespace ::model;
+
+void MetricBasedDomain::ThrowIfEmpty() const {
+    if (metrizable_types_.empty()) {
+        throw std::runtime_error(
+                "Metric-based domain must be instantiated with types (SetTypes) before use");
+    }
+}
+
+bool MetricBasedDomain::Compare(Tuple const& x, Tuple const& y) const {
+    ThrowIfEmpty();
+    assert(x.size() >= metrizable_types_.size());
+    assert(y.size() >= metrizable_types_.size());
+
+    return CompareInternal(x, y);
+}
+
+double MetricBasedDomain::DistBetweenBytes(std::size_t type_num, std::byte const* x,
+                                           std::byte const* y) const {
+    if (x == nullptr || y == nullptr) {
+        return dist_from_null_is_infty_ ? std::numeric_limits<double>::infinity() : 0;
+    }
+    return metrizable_types_[type_num]->Dist(x, y) * leveling_coeffs_[type_num];
+}
+
+std::vector<std::byte const*> MetricBasedDomain::AllocateValues(
+        std::vector<std::string> const& str_values) const {
+    assert(str_values.size() >= metrizable_types_.size());
```

I suppose that it's better to throw an exception instead of writing an assert because both sizes depend on user's input (`str_values` are obtained from the domain and `metrizable_types_` are obtained from `column_indices` passed to `load_data`). I also guess that `==` instead of `>=` should be used.